core/src/test/java/org/apache/lucene/search/uhighlight/BoundedBreakIteratorScannerTests.java

```diff
@@ -0,0 +1,130 @@
+                String term = vocabulary[termId].toLowerCase(Locale.ROOT);
+                if (j == 0) {
+                    // capitalize the first letter of the first term in the sentence
+                    term = term.substring(0, 1).toUpperCase(Locale.ROOT) + term.substring(1);
+                } else {
+                    // word boundary
+                    String sep = randomFrom(" ", "  ",  "\t", "#", "\n");
+                    text.append(sep);
+                }
+                maxTermLen = Math.max(term.length(), maxTermLen);
+                offsetList.add(text.length());
+                sizeList.add(term.length());
+                text.append(term);
+            }
+            // sentence boundary
+            String boundary = randomFrom("! ", "? ", ". ", ".\n", ".\n\n");
+            text.append(boundary);
+        }
+
+        int[] sizes = sizeList.stream().mapToInt(i->i).toArray();
+        int[] offsets = offsetList.stream().mapToInt(i->i).toArray();
+
+        bi.setText(text.toString());
+        int currentPos = randomIntBetween(0, 20);
+        int lastEnd = -1;
+        int maxPassageLen = maxLen+(maxTermLen*2);
+        while (currentPos < offsets.length) {
+            // find the passage that contains the current term
+            int nextOffset = offsets[currentPos];
+            int start = bi.preceding(nextOffset+1);
+            int end = bi.following(nextOffset);
+
+            // check that the passage is valid
+            assertThat(start, greaterThanOrEqualTo(lastEnd));
+            assertThat(end, greaterThan(start));
+            assertThat(start, lessThanOrEqualTo(nextOffset));
+            assertThat(end, greaterThanOrEqualTo(nextOffset));
+            int passageLen = end-start;
+            assertThat(passageLen, lessThanOrEqualTo(maxPassageLen));
+
+            // checks that the start and end of the passage are on word boundaries.
+            int startPos = Arrays.binarySearch(offsets, start);
+            int endPos = Arrays.binarySearch(offsets, end);
+            if (startPos < 0) {
+                int lastWordEnd = offsets[Math.abs(startPos)-2] + sizes[Math.abs(startPos)-2];
+                assertThat(start, greaterThanOrEqualTo(lastWordEnd));
+            }
+            if (endPos < 0) {
+                if (Math.abs(endPos)-2 < offsets.length) {
+                    int lastWordEnd = offsets[Math.abs(endPos) - 2] + sizes[Math.abs(endPos) - 2];
+                    assertThat(end, greaterThanOrEqualTo(lastWordEnd));
+                }
+                // advance the position to the end of the current passage
+                currentPos = (Math.abs(endPos) - 1);
+            } else {
+                // advance the position to the end of the current passage
+                currentPos = endPos;
+            }
+            // randomly advance to the next term to highlight
+            currentPos += randomIntBetween(0, 20);
+            lastEnd = end;
+        }
+    }
+
```

Comment:
Can you add a non-random test just so it is easier to see? One with a "normal" sentence and maybe one with a long one? Just for illustration purposes.